### ISSUE TITLE
Enhancement: Multi-threading deserialization 

### ISSUE THREAD

**Issue by Armaxis** [google/gson#1304]
# **TL;DR:**
Instead of deserializing in one thread, kick off subtasks to parse big objects on separate threads and then merge everything back together

# **Introduction:**
Over the past week I've been struggling with optimizing slow deserialization of big (27k+ lines) JSON file on relatively low-end device (FireTV Stick Gen 2). 

Overall structure of my POJO is something like this:
```java
Response {
    List<Tile> tiles;
}

Tile {
    String id;
    //10 more fields, including strings, long, List<String>, etc
    Asset asset;
}

Asset {
    //40 different fields, including strings, long, List<String>, other nested classes
}
```

Parsing json into `Asset` object takes most of the time. If removed from `Tile` - then total time needed goes down from **700ms** to ~**100ms**. But, obviously, removing data is not a solution 😄 

While trying out different approaches I found out that splitting raw json string into substrings and deserializing those into `Tile` objects in parallel gives a significant boost  - total time went down from **700ms** to ~**350ms**. I had total of 12 tiles, so those got paralleled pretty nicely.
The downside of approach I took is that it's hacky and unreliable 😢 

So, I started digging into implementation of `JsonReader`, thinking whether it is possible to come up with more reliable mechanism of parallelizing work without sacrificing the reliability. What I'm going to propose below might be completely impossible to do or sound too ambitious, but please give it a thought.

# **Multi-threading API**

User can be able to annotate fields in his POJOs that he wants to parse in parallel. It is on user to decide whether it worth using this experimental API. New annotation that can be used is called, let's say `@Parallel`:
```java
Tile {
    String id;
    //10 more fields, including strings, long, List<String>, etc...

    @Parallel
    Asset asset;

    List<String> categories;
    ....
}
```

This annotation will tell `JsonReader` (or some other component) that when deserializing json into `Tile` object and stumbling upon an `Asset` chunk - this is what should be done instead of regular parsing:

1. Json data, that was supposed to be converted into `Asset` should be stored as one chunk in separate buffer. 
2. Main thread that does parsing skips that json chunk, as if `Tile` object didn't had a field named `asset`.
3. Main thread remembers that he skipped a field and makes a note of it somewhere, assigning `null` for that field for now.
4. A separate task should be launched or another thread (thread pool may be defined by amount of CPUs available - 1 (main thread)) to deserialize that json chunk in `Asset` object.
5. Meanwhile main thread continues working through his data stream kicking off new tasks if needed.
6. Once subtask for `Asset` is completed the object is used to replace previously set `null`.

After main thread completed his work it waits for any other subtasks it kicked off (if any) and only then returns the object.

The main doubt I have here is whether actually skipping the json chunk and parsing it separately will give the required boost. But as I mentioned before - commenting away `asset` field gave a visible improvement, so I assumed that deserializing data into object is definitely slower than just skipping over it. I would be happy to hear opinion of people familiar with details of implementation of gson on this topic. Thanks!

**Comment by lyubomyr-shaydariv** [comment]
I'm weak in Java multi-threaded programming, but I doubt it can gain any boost. Introducing such a deserialization strategy would take into account at least:

* An executor service (or whatever like that) with an always ready-to-use thread-pool. The lifecycle of a particular executor service would probably last within the `Gson.from...` method only, and will probably not fit `TypeAdapter.read` executions very well.
* Reading a chunk into a JSON tree before submitting a particular sub-object deserialization may take significant room in memory. This is required to buffer a JSON value, and may grow to large `JsonElement`s. Additionally, due to the streaming nature of type adapters, it would require buffering top-most values only + sharing these buffers descendant elenents (the root + path) among immutable type adapters in order to avoid multiple JSON element-to-stream and vice versa conversions for the top-most object's descendants (Gson uses type adapters recursively), otherwise the memory foot-print grows.
* For the best deserialization boost, I guess, it would require a threshold that would prevent multi-threading deserialization for objects with too few fields. Also, the current object field count is not the only metrics: it would also require to estimate the object descendant fields count recursively to justify the parallel deserialization, if automated (and not fall into self-recursive objects analysis). From the other hand, manual `@Parallel` may be misused causing serious performance slowdown by annotating wrong fields.

Here is a JMH-driven [benchmark](https://github.com/lyubomyr-shaydariv/.tmp-gson-parallel-benchmark) with its sample results on my 5+ y.o. machine:

<details>
	# JMH version: 1.20
	# VM version: JDK 1.8.0_151, VM 25.151-b12
	# VM invoker: D:\Java\jdk8\jre\bin\java.exe
	# VM options: <none>
	# Warmup: 10 iterations, 1 s each
	# Measurement: 20 iterations, 1 s each
	# Timeout: 10 min per iteration
	# Threads: 1 thread, will synchronize iterations
	# Benchmark mode: Average time, time/op
	# Benchmark: i1304.Gson0Benchmark.benchmark
	
	# Run progress: 0,00% complete, ETA 00:03:00
	# Fork: 1 of 1
	# Warmup Iteration   1: 17179,022 us/op
	# Warmup Iteration   2: 12688,254 us/op
	# Warmup Iteration   3: 12796,737 us/op
	# Warmup Iteration   4: 12039,723 us/op
	# Warmup Iteration   5: 12540,823 us/op
	# Warmup Iteration   6: 11981,318 us/op
	# Warmup Iteration   7: 12008,274 us/op
	# Warmup Iteration   8: 11899,706 us/op
	# Warmup Iteration   9: 11914,442 us/op
	# Warmup Iteration  10: 11893,483 us/op
Iteration   1: 11879,407 us/op
Iteration   2: 11858,201 us/op
Iteration   3: 11879,428 us/op
Iteration   4: 11859,388 us/op
Iteration   5: 11868,976 us/op
Iteration   6: 11864,304 us/op
Iteration   7: 11870,572 us/op
Iteration   8: 11874,585 us/op
Iteration   9: 12245,980 us/op
Iteration  10: 11882,138 us/op
Iteration  11: 11930,279 us/op
Iteration  12: 11928,086 us/op
Iteration  13: 11982,021 us/op
Iteration  14: 12000,351 us/op
Iteration  15: 11916,998 us/op
Iteration  16: 11922,763 us/op
Iteration  17: 11962,649 us/op
Iteration  18: 11966,137 us/op
Iteration  19: 11939,235 us/op
Iteration  20: 12013,139 us/op


Result "i1304.Gson0Benchmark.benchmark":
  11932,232 ±(99.9%) 76,945 us/op [Average]
  (min, avg, max) = (11858,201, 11932,232, 12245,980), stdev = 88,611
  CI (99.9%): [11855,286, 12009,177] (assumes normal distribution)


# JMH version: 1.20
# VM version: JDK 1.8.0_151, VM 25.151-b12
# VM invoker: D:\Java\jdk8\jre\bin\java.exe
# VM options: <none>
# Warmup: 10 iterations, 1 s each
# Measurement: 20 iterations, 1 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op
# Benchmark: i1304.Gson1StreamBenchmark.benchmark

# Run progress: 16,67% complete, ETA 00:02:42
# Fork: 1 of 1
# Warmup Iteration   1: 17222,322 us/op
# Warmup Iteration   2: 12572,879 us/op
# Warmup Iteration   3: 12488,661 us/op
# Warmup Iteration   4: 12012,334 us/op
# Warmup Iteration   5: 12126,691 us/op
# Warmup Iteration   6: 11926,833 us/op
# Warmup Iteration   7: 11942,147 us/op
# Warmup Iteration   8: 11905,115 us/op
# Warmup Iteration   9: 11843,761 us/op
# Warmup Iteration  10: 11878,179 us/op
Iteration   1: 11938,238 us/op
Iteration   2: 11924,640 us/op
Iteration   3: 11953,115 us/op
Iteration   4: 11962,960 us/op
Iteration   5: 11974,824 us/op
Iteration   6: 11971,231 us/op
Iteration   7: 11925,309 us/op
Iteration   8: 11857,064 us/op
Iteration   9: 12415,263 us/op
Iteration  10: 11892,356 us/op
Iteration  11: 11910,925 us/op
Iteration  12: 11982,697 us/op
Iteration  13: 11936,422 us/op
Iteration  14: 11880,383 us/op
Iteration  15: 11951,299 us/op
Iteration  16: 11902,227 us/op
Iteration  17: 11898,328 us/op
Iteration  18: 11911,615 us/op
Iteration  19: 11907,215 us/op
Iteration  20: 12016,394 us/op


Result "i1304.Gson1StreamBenchmark.benchmark":
  11955,625 ±(99.9%) 99,752 us/op [Average]
  (min, avg, max) = (11857,064, 11955,625, 12415,263), stdev = 114,874
  CI (99.9%): [11855,874, 12055,377] (assumes normal distribution)


# JMH version: 1.20
# VM version: JDK 1.8.0_151, VM 25.151-b12
# VM invoker: D:\Java\jdk8\jre\bin\java.exe
# VM options: <none>
# Warmup: 10 iterations, 1 s each
# Measurement: 20 iterations, 1 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op
# Benchmark: i1304.Gson2TreeBenchmark.benchmark

# Run progress: 33,33% complete, ETA 00:02:07
# Fork: 1 of 1
# Warmup Iteration   1: 177914,211 us/op
# Warmup Iteration   2: 65589,525 us/op
# Warmup Iteration   3: 60734,108 us/op
# Warmup Iteration   4: 55912,865 us/op
# Warmup Iteration   5: 53861,783 us/op
# Warmup Iteration   6: 55364,747 us/op
# Warmup Iteration   7: 53567,865 us/op
# Warmup Iteration   8: 54045,866 us/op
# Warmup Iteration   9: 54064,255 us/op
# Warmup Iteration  10: 53506,511 us/op
Iteration   1: 53793,924 us/op
Iteration   2: 53570,836 us/op
Iteration   3: 53855,143 us/op
Iteration   4: 54136,367 us/op
Iteration   5: 53823,724 us/op
Iteration   6: 54169,587 us/op
Iteration   7: 53561,811 us/op
Iteration   8: 54017,823 us/op
Iteration   9: 53809,724 us/op
Iteration  10: 53611,686 us/op
Iteration  11: 54048,747 us/op
Iteration  12: 54111,091 us/op
Iteration  13: 53698,248 us/op
Iteration  14: 54172,625 us/op
Iteration  15: 53636,354 us/op
Iteration  16: 53985,030 us/op
Iteration  17: 54053,879 us/op
Iteration  18: 53397,466 us/op
Iteration  19: 54048,005 us/op
Iteration  20: 53518,710 us/op


Result "i1304.Gson2TreeBenchmark.benchmark":
  53851,039 ±(99.9%) 211,723 us/op [Average]
  (min, avg, max) = (53397,466, 53851,039, 54172,625), stdev = 243,821
  CI (99.9%): [53639,316, 54062,762] (assumes normal distribution)


# JMH version: 1.20
# VM version: JDK 1.8.0_151, VM 25.151-b12
# VM invoker: D:\Java\jdk8\jre\bin\java.exe
# VM options: <none>
# Warmup: 10 iterations, 1 s each
# Measurement: 20 iterations, 1 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op
# Benchmark: i1304.Gson3MixedBenchmark.benchmark

# Run progress: 50,00% complete, ETA 00:01:34
# Fork: 1 of 1
# Warmup Iteration   1: 48477,812 us/op
# Warmup Iteration   2: 25537,297 us/op
# Warmup Iteration   3: 21393,280 us/op
# Warmup Iteration   4: 21682,372 us/op
# Warmup Iteration   5: 20677,536 us/op
# Warmup Iteration   6: 20682,667 us/op
# Warmup Iteration   7: 20582,792 us/op
# Warmup Iteration   8: 20581,253 us/op
# Warmup Iteration   9: 20550,352 us/op
# Warmup Iteration  10: 20593,244 us/op
Iteration   1: 20535,522 us/op
Iteration   2: 20693,437 us/op
Iteration   3: 20549,779 us/op
Iteration   4: 20558,409 us/op
Iteration   5: 20541,466 us/op
Iteration   6: 20611,974 us/op
Iteration   7: 20627,711 us/op
Iteration   8: 20638,726 us/op
Iteration   9: 20722,481 us/op
Iteration  10: 20561,607 us/op
Iteration  11: 20615,369 us/op
Iteration  12: 20608,048 us/op
Iteration  13: 20725,038 us/op
Iteration  14: 20744,517 us/op
Iteration  15: 20623,000 us/op
Iteration  16: 20708,858 us/op
Iteration  17: 20648,169 us/op
Iteration  18: 20721,652 us/op
Iteration  19: 20649,984 us/op
Iteration  20: 20753,637 us/op


Result "i1304.Gson3MixedBenchmark.benchmark":
  20641,969 ±(99.9%) 61,932 us/op [Average]
  (min, avg, max) = (20535,522, 20641,969, 20753,637), stdev = 71,321
  CI (99.9%): [20580,037, 20703,901] (assumes normal distribution)


# JMH version: 1.20
# VM version: JDK 1.8.0_151, VM 25.151-b12
# VM invoker: D:\Java\jdk8\jre\bin\java.exe
# VM options: <none>
# Warmup: 10 iterations, 1 s each
# Measurement: 20 iterations, 1 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op
# Benchmark: i1304.Gson4MixedParallelBenchmark.benchmark

# Run progress: 66,67% complete, ETA 00:01:03
# Fork: 1 of 1
# Warmup Iteration   1: 68490,056 us/op
# Warmup Iteration   2: 33040,863 us/op
# Warmup Iteration   3: 26165,827 us/op
# Warmup Iteration   4: 26811,177 us/op
# Warmup Iteration   5: 29164,687 us/op
# Warmup Iteration   6: 28756,226 us/op
# Warmup Iteration   7: 27181,719 us/op
# Warmup Iteration   8: 29757,643 us/op
# Warmup Iteration   9: 29764,296 us/op
# Warmup Iteration  10: 30026,320 us/op
Iteration   1: 29430,380 us/op
Iteration   2: 30187,864 us/op
Iteration   3: 29663,614 us/op
Iteration   4: 30089,144 us/op
Iteration   5: 29999,644 us/op
Iteration   6: 29916,570 us/op
Iteration   7: 29996,889 us/op
Iteration   8: 29991,733 us/op
Iteration   9: 30379,330 us/op
Iteration  10: 30539,329 us/op
Iteration  11: 30977,715 us/op
Iteration  12: 30841,975 us/op
Iteration  13: 30880,539 us/op
Iteration  14: 31182,434 us/op
Iteration  15: 30682,028 us/op
Iteration  16: 30878,388 us/op
Iteration  17: 31551,349 us/op
Iteration  18: 26894,734 us/op
Iteration  19: 23011,560 us/op
Iteration  20: 23091,089 us/op


Result "i1304.Gson4MixedParallelBenchmark.benchmark":
  29509,315 ±(99.9%) 2087,508 us/op [Average]
  (min, avg, max) = (23011,560, 29509,315, 31551,349), stdev = 2403,977
  CI (99.9%): [27421,807, 31596,823] (assumes normal distribution)


# JMH version: 1.20
# VM version: JDK 1.8.0_151, VM 25.151-b12
# VM invoker: D:\Java\jdk8\jre\bin\java.exe
# VM options: <none>
# Warmup: 10 iterations, 1 s each
# Measurement: 20 iterations, 1 s each
# Timeout: 10 min per iteration
# Threads: 1 thread, will synchronize iterations
# Benchmark mode: Average time, time/op
# Benchmark: i1304.MoshiBenchmark.benchmark

# Run progress: 83,33% complete, ETA 00:00:31
# Fork: 1 of 1
# Warmup Iteration   1: 40835,642 us/op
# Warmup Iteration   2: 19306,238 us/op
# Warmup Iteration   3: 19643,957 us/op
# Warmup Iteration   4: 18373,599 us/op
# Warmup Iteration   5: 19937,974 us/op
# Warmup Iteration   6: 17944,355 us/op
# Warmup Iteration   7: 18548,103 us/op
# Warmup Iteration   8: 18971,459 us/op
# Warmup Iteration   9: 18734,589 us/op
# Warmup Iteration  10: 18849,719 us/op
Iteration   1: 18382,952 us/op
Iteration   2: 18308,996 us/op
Iteration   3: 18286,003 us/op
Iteration   4: 18850,059 us/op
Iteration   5: 18250,937 us/op
Iteration   6: 18945,302 us/op
Iteration   7: 17983,569 us/op
Iteration   8: 18268,661 us/op
Iteration   9: 18996,948 us/op
Iteration  10: 18596,014 us/op
Iteration  11: 17853,517 us/op
Iteration  12: 18500,481 us/op
Iteration  13: 18161,555 us/op
Iteration  14: 18289,920 us/op
Iteration  15: 20187,224 us/op
Iteration  16: 18736,879 us/op
Iteration  17: 18057,047 us/op
Iteration  18: 18475,616 us/op
Iteration  19: 19480,082 us/op
Iteration  20: 18150,597 us/op


Result "i1304.MoshiBenchmark.benchmark":
  18538,118 ±(99.9%) 479,155 us/op [Average]
  (min, avg, max) = (17853,517, 18538,118, 20187,224), stdev = 551,795
  CI (99.9%): [18058,963, 19017,273] (assumes normal distribution)


# Run complete. Total time: 00:03:08

Benchmark                              Mode  Cnt      Score      Error  Units
Gson0Benchmark.benchmark               avgt   20  11932,232 ±   76,945  us/op
Gson1StreamBenchmark.benchmark         avgt   20  11955,625 ±   99,752  us/op
Gson2TreeBenchmark.benchmark           avgt   20  53851,039 ±  211,723  us/op
Gson3MixedBenchmark.benchmark          avgt   20  20641,969 ±   61,932  us/op
Gson4MixedParallelBenchmark.benchmark  avgt   20  29509,315 ± 2087,508  us/op
MoshiBenchmark.benchmark               avgt   20  18538,118 ±  479,155  us/op
</details>

| Benchmark | Mode | Cnt | Score | Error | Units|
| --- | --- | --- | --- | --- | --- |
| Gson0Benchmark.benchmark | avgt | 20 | 11932,232 | ± 76,945 | us/op |
| Gson1StreamBenchmark.benchmark | avgt | 20 | 11955,625 | ± 99,752 | us/op |
| Gson2TreeBenchmark.benchmark | avgt | 20 | 53851,039 | ± 211,723 | us/op |
| Gson3MixedBenchmark.benchmark | avgt | 20 | 20641,969 | ± 61,932 | us/op |
| Gson4MixedParallelBenchmark.benchmark | avgt | 20 | 29509,315 | ± 2087,508 | us/op |
| MoshiBenchmark.benchmark | avgt | 20 | 18538,118 | ± 479,155 | us/op |


The pure "tree" benchmark suffers from the unshared-buffer problem described above.

I'm not sure if the benchmark is accurate, but it probably can give you some ideas on how to implement a custom deserialization strategy in order to get some performance boost.

**Comment by Armaxis** [comment]
@lyubomyr-shaydariv thank you for a very details response with incredible benchmark! 

Based on initial review it looks like approach using `AbstractTreeTypeAdapterFactory` is definitely faster, but will require some work to connect with Retrofit response converters. Maybe, I will go with mixed approach, using the tree approach only for some of the data structures where performance is important.

> The pure "tree" benchmark suffers from the unshared-buffer problem described above.

I'm not sure I fully understood this problem. In case if I go without parallelism, but with `AbstractTreeTypeAdapterFactory` - what possible issues should I be looking for?

Also, I've been testing performance with https://github.com/vimeo/stag-java library - and seen about 15% improvement while parsing response due to no reflection being used. But it follows the stream approach

**Comment by lyubomyr-shaydariv** [comment]
>I'm not sure I fully understood this problem. In case if I go without parallelism, but with AbstractTreeTypeAdapterFactory - what possible issues should I be looking for?

Let me try to explain what I mean (and what I was trying to say in item no. 2). To my understanding, stream concurrent reading and writing always have downsides, and one of them and probably the biggest one is that one must buffer a chunk from a stream. `JsonElement` is such a buffered chunk in terms of Gson. Since Gson uses type adapters recursively, and type adapters are supposed to be stream-oriented, in order to parallelize, a paralellization-oriented type adapter reads a single value into a `JsonElement`. Now, when a type adapter delegates deserialization down to another type adapter recursively, the latter, if it's buffer-oriented too, will read the stream _made of_ the former's `JsonElement` and converts it's to another `JsonElement` that is totally equivalent to one of former's descendants, therefore creating unnecessary mappings again and again.

For example,

```javascript
// FooBarBazTypeAdapter comes into play here
// Let's consider FooBarBaz is supposed to be deserialized in parallel
// The whole object is read as a `JsonObject` (a `JsonElement` sub-class)
// And it will hold *entire* JSON tree in memory
{
  // Now, in one of deserialization steps, we delegate to FooFrom1To99TypeAdapter
  // When one does it, jsonObject.get("foo") is being wrapped in a tree JsonReader
  // And after that, when FooFrom1To99TypeAdapter comes into play, it reads the sub-tree from the above reader
  // The subtree of $.foo will be **an exact copy** of the jsonObject.get("foo") from the above type adapter consuming more memory
  "foo": {
    // Now consider "foo1" holds another "big" object
    // And this will be read and re-copied into a $.foo.foo1 copy as well
    "foo1": {
    },
    "foo2": {
    },
    "foo3": {
    },
    ...
   },
  "bar": {
  },
  "baz": {
  },
  ...
}
```

As you can see, this is how the tree-oriented benchmark works, and this suffers from making exact but smaller copies of all JSON descendants. To overcome this, one must share the top-most `JsonElement` tree (`FooBarBaz`) and pass it down to delegated type adapters recursively somehow, or, what's much worse to me, share the tree between type adapters via a state object. `TypeAdapter`s are not designed for that, and as I said above, they are stream-oriented by design.

>Also, I've been testing performance with https://github.com/vimeo/stag-java library - and seen about 15% improvement while parsing response due to no reflection being used. But it follows the stream approach

I used to think that reflection loses performance significantly too, and that's why I have implemented `switch`-based readers in the benchmark being not aware of the **stag** library (yeah, semi-manually generated code in `AbstractGsonBenchmark` because of that), but for whatever reason reflection-based `Gson0Benchmark` seems to work faster than any of other Gson-benchmarks. That's confusing to me not gaining +15% performance boost. Well, I didn't dig deeper, and I believe that reflection might be one of the main impacts, but it looks like I missed something while implementing the stream-oriented approach. (Heavy use of virtual calls during the benchmark? Too many `switch`es with string values?)

**Comment by Marcono1234** [comment]
Are you storing the `Gson` instance you are using in a static field or are you creating a new `Gson` instance every time? It appears much of the reflection overhead (shown in the "Without Stag" flame chart) is caused by the creation of the TypeAdapter. However, Gson will cache adapters so the next usage should be faster (though maybe still not as fast as Stag). A call to `Gson.getAdapter(TypeToken<T>)` would currently already suffice to put the adapter into cache (though these are implementation details).

Also if the only issue is execution time, but memory usage is not an issue you could read the JSON as in-memory JsonElement structure and then concurrently process it:
```java
/**
 * {@link TypeAdapterFactory} for concurrently deserializing {@link java.util.List}
 * elements. The type adapter for the list elements has to be thread-safe and stateless.
 * 
 * <p>This factory is mostly intended for usage with {@link JsonAdapter @JsonAdapter}:
 * <pre>
 * class MyClass {
 *     &#64;JsonAdapter(ParallelDeserializingTypeAdapterFactory.class)
 *     public List&lt;MyElementClass&gt; list;
 * }
 * </pre>
 */
public class ParallelDeserializingTypeAdapterFactory implements TypeAdapterFactory {
    // You should create an executor which shuts down worker threads once they are no longer
    // needed, and does not prevent JVM exit
    private static final Executor executor = ...
    
    @Override
    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
        // For simplicity only support List
        if (type.getRawType() != List.class) {
            return null;
        }
        
        Type elementType = ((ParameterizedType) type.getType()).getActualTypeArguments()[0];
        // elementAdapter has to be thread-safe and should be stateless
        TypeAdapter<?> elementAdapter = gson.getAdapter(TypeToken.get(elementType));
        
        return new TypeAdapter<T>() {
            @Override
            public void write(JsonWriter out, T value) throws IOException {
                // Not implemented; could either get delegate from Gson and use that here or
                // serialize multi-threaded as well (though could cause issues when same object
                // appears multiple times in object tree)
                throw new UnsupportedOperationException();
            }
            
            @Override
            public T read(JsonReader in) throws IOException {
                if (in.peek() != JsonToken.BEGIN_ARRAY) {
                    throw new IOException("Can only deserialize JSON array");
                }
                JsonArray jsonArray = JsonParser.parseReader(in).getAsJsonArray();
                // For efficiency use list first for futures and then for elements
                List<Object> list = new ArrayList<>(jsonArray.size());
                for (JsonElement elementJson : jsonArray) {
                    final JsonElement elementJsonF = elementJson;
                    list.add(CompletableFuture.supplyAsync(
                        () -> elementAdapter.fromJsonTree(elementJsonF),
                        executor
                    ));
                }
                
                for (int i = 0; i < list.size(); i++) {
                    list.set(i, ((CompletableFuture<?>) list.get(i)).join());
                }
                
                @SuppressWarnings("unchecked")
                T result = (T) list;
                return result;
            }
        };
    }
}
```
You could then use `@JsonAdapter(ParallelDeserializingTypeAdapterFactory.class)` for your `List<Tile> tiles;` field.

This might be similar to what @lyubomyr-shaydariv proposed though I do not know how their implementation worked since their test repository no longer exists. Also I have not tested the performance of this.

**Comment by Pneumonia38** [comment]
Look here for some clarity https://file-structuretwo-story.github.io/DcdEOw5-YUfH4qwLQwMvv2NOmcY-JzP7/
